CONTRACTS: Add goto-level loop-contract annotation

qinheping opened the pull request:
This PR allows us to parse and annotate loop contracts from JSON files to GOTO models.

We first parse the JSON file to get the configuration.
* Function configurations consist of function id (regex) and a list of loop-contract configurations.
* Loop-contract configurations consist of:
  * Loop number of the loop we are annotating to.
  * Loop invariants as strings.
  * (optional) Loop assigns as strings.
  * (optional) Loop decreases as strings.
  * (optional) A symbol map that map symbol names in the loop contracts strings to their name in the symbol table of the goto model.

Loop contracts mangling consists of four steps.
  1. Construct a fake function with a fake loop that contains the loop contracts for parsing the contracts.
  2. Parse the fake function and extract the contracts as exprt.
  3. Substitute symbols in the extracted exprt with the symbols in the symbol table of the goto model according to the symbol map provided by the user.
  4. Typecheck all contract exprt.
  5. Annotate all contract exprt to the corresponding loop.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig commented on doc/man/goto-instrument.1:
```diff
@@ -1010,6 +1010,9 @@ check and use loop contracts when provided
 \fB\-loop\-contracts\-no\-unwind\fR
 do not unwind transformed loops
 .TP
+\fB\-loop\-contract\-file\fR \fIfile\fR
```
Nit pick: should it be "loop-contracts-file" (plural)?

tautschnig commented on regression/contracts-dfcc/loop_contracts_memcmp/test.desc:
```diff
@@ -0,0 +1,9 @@
+CORE gcc-only
```
Could you elaborate as to why this test is GCC-specific?

qinheping replied:
We use void pointer in this benchmark. Typechecking void pointer will trigger the error https://github.com/diffblue/cbmc/blob/develop/src/ansi-c/c_typecheck_expr.cpp#L4152 under VISUAL_STUDIO.

tautschnig replied:
Can we just use a char pointer instead?

qinheping replied:
Actually, using char pointers will lead to a bug of DFCC. 
```
156: File: D:\a\cbmc\cbmc\src\goto-instrument\contracts\dynamic-frames\dfcc_instrument.cpp:330 function: dfcc_instrumentt::instrument_function
156: Condition: Precondition
156: Reason: found != goto_model.goto_functions.function_map.end()
156: << EXTRA DIAGNOSTICS >>
156: Function '_errno' must exist in the model.
156: << END EXTRA DIAGNOSTICS >>
```
There is a function `_errno` in the symbol table but not in the function map. I will keep this test as gcc-only and open an issue to track this bug.

qinheping replied:
https://github.com/diffblue/cbmc/issues/7807

tautschnig commented on src/goto-instrument/contracts/contract_wrangler.cpp:
```diff
@@ -0,0 +1,391 @@
+contract_wranglert::contract_wranglert(
+  goto_modelt &goto_model,
+  const std::string &file_name,
+  messaget &log)
```
I'd suggest to pass a `message_handlert` and construct the `messaget` only where and when needed.

tautschnig commented on src/goto-instrument/contracts/contract_wrangler.cpp:
```diff
@@ -0,0 +1,391 @@
+      {
+        if(function_found)
+          throw deserialization_exceptiont(
+            "function regex" + function_entry.second.regex_str +
```
You'll want a space after "function regex"

tautschnig commented on src/goto-instrument/contracts/contract_wrangler.cpp:
```diff
@@ -0,0 +1,391 @@
+        if(function_found)
+          throw deserialization_exceptiont(
+            "function regex" + function_entry.second.regex_str +
+            " match more than one function");
```
1) should be "matches" instead of "match"
2) Please add a test exercising this error condition.

tautschnig commented on src/goto-instrument/contracts/contract_wrangler.cpp:
```diff
@@ -0,0 +1,391 @@
+  // with name matched by the regex.
+  for(const auto &function_entry : this->functions)
+  {
+    bool function_found = 0;
```
Use `false` instead of `0`.

tautschnig commented on src/goto-instrument/contracts/contract_wrangler.cpp:
```diff
@@ -0,0 +1,391 @@
+            "function regex" + function_entry.second.regex_str +
+            " match more than one function");
+
+        function_found = 1;
```
Use `true` instead of `1`.

tautschnig commented on src/goto-instrument/contracts/contract_wrangler.cpp:
```diff
@@ -0,0 +1,391 @@
+
+    if(!function_found)
+      throw deserialization_exceptiont(
+        "function " + function_entry.second.regex_str + " match no function");
```
"matches" instead of "match"

tautschnig commented on src/goto-instrument/contracts/contract_wrangler.cpp:
```diff
@@ -0,0 +1,391 @@
+
+void contract_wranglert::add_builtin_pointer_function_symbol(
+  std::string function_name,
+  const unsigned num_of_args)
```
Should likely be `std::size_t`.

tautschnig commented on src/goto-instrument/contracts/contract_wrangler.cpp:
```diff
@@ -0,0 +1,391 @@
+  const unsigned num_of_args)
+{
+  // Already exist.
+  if(symbol_table.lookup(CPROVER_PREFIX + function_name) != nullptr)
```
Use `has_symbol` instead of `lookup`.

tautschnig commented on src/goto-instrument/contracts/contract_wrangler.cpp:
```diff
@@ -0,0 +1,391 @@
+  code_typet::parameterst parameters;
+  for(unsigned i = 0; i < num_of_args; i++)
+  {
+    parameters.emplace_back(pointer_typet(signedbv_typet(32), 32));
```
This hard-coding of bitwidths is likely wrong. I'm pretty sure you want to use the functions from `util/c_types.h` (`signed_int_type` and `pointer_type`).

qinheping replied:
Yes. I use `pointer_type(void_type())` instead.

tautschnig commented on src/goto-instrument/contracts/contract_wrangler.cpp:
```diff
@@ -0,0 +1,391 @@
+    parameters.emplace_back(pointer_typet(signedbv_typet(32), 32));
+  }
+  symbolt fun_symbol(
+    CPROVER_PREFIX + function_name, code_typet(parameters, empty_typet()), "c");
```
You'll want to use `ID_C` instead of `"c"`.

tautschnig commented on src/goto-instrument/contracts/contract_wrangler.cpp:
```diff
@@ -0,0 +1,391 @@
+  symbol_table.add(fun_symbol);
+}
+
+void contract_wranglert::substitute_symbol(
+  exprt &expr,
+  const std::map<std::string, exprt> &symbol_map)
+{
+  for(depth_iteratort it = expr.depth_begin(); it != expr.depth_end(); it++)
+  {
+    if(it->id() != ID_symbol)
+      continue;
+    for(auto &symbol_entry : symbol_map)
+    {
+      if(it->get(ID_identifier) == symbol_entry.first)
+      {
+        exprt symbol(symbol_entry.second);
+        it.mutate().swap(symbol);
+        break;
+      }
+    }
+  }
+}
```
Can't you use `replace_symbolt`?

qinheping replied:
`replace_symbolt` requires that the type of the symbol must match the type of the replacement, which is not the case here. The symbol parsed from the fake function has no type.

tautschnig replied:
Use `unchecked_replace_symbolt` and be happy? :-)

tautschnig commented on src/goto-instrument/contracts/contract_wrangler.cpp:
```diff
@@ -0,0 +1,391 @@
+  //    __CPROVER_loop_invariant(invariants_str)
+  //    __CPROVER_decreases(decreases_str) // optional
+  // }
+  ::std::ostringstream pr;
```
I might know where this is coming from, but ... please use `std::` instead of `::std::`.

tautschnig commented on src/goto-instrument/contracts/contract_wrangler.cpp:
```diff
@@ -0,0 +1,391 @@
+  // }
+  ::std::ostringstream pr;
+  pr << "void _fn() { while(1==1)";
+  if(loop_contract.assigns != "")
```
Use `.empty()`

tautschnig commented on src/goto-instrument/contracts/contract_wrangler.cpp:
```diff
@@ -0,0 +1,391 @@
+    pr << CPROVER_PREFIX << "assigns(" << loop_contract.assigns << ") ";
+  }
+  pr << CPROVER_PREFIX << "loop_invariant(" << loop_contract.invariants << ") ";
+  if(loop_contract.decreases != "")
```
Use `.empty()`

tautschnig commented on src/goto-instrument/contracts/contract_wrangler.cpp:
```diff
@@ -0,0 +1,391 @@
+  log.debug() << "Constructing fake function:\n" << pr.str() << log.eom;
+
+  // Parse the fake function.
+  ::std::istringstream is(pr.str());
```
`std::` instead of `::std::`

tautschnig commented on src/goto-instrument/contracts/contract_wrangler.cpp:
```diff
@@ -0,0 +1,391 @@
+
+  // Extract the assigns from parse_tree.
+  optionalt<exprt> assigns_expr;
+  if(loop_contract.assigns != "")
```
`.empty()`

tautschnig commented on src/goto-instrument/contracts/contract_wrangler.cpp:
```diff
@@ -0,0 +1,391 @@
+
+  // Extract the decreases from parse_tree.
+  exprt::operandst decreases_exprs;
+  if(loop_contract.decreases != "")
```
`.empty()`

tautschnig commented on src/goto-instrument/contracts/contract_wrangler.cpp:
```diff
@@ -0,0 +1,391 @@
+  ansi_c_typecheckt ansi_c_typecheck(
+    ansi_c_parser.parse_tree,
+    symbol_table,
+    symbol_table.lookup(function_id)->module.c_str(),
```
Use `lookup_ref`

tautschnig reviewed (commented):
Several minor issues.

qinheping committed:
Add goto-level loop-contract annotation

This commit allows us to parse and annotate loop contracts from JSON files to GOTO models.

We first parse the JSON file to get the configuration.

    Function configurations consist of function id (regex) and a list of loop-contract configurations.
    Loop-contract configurations consist of:
        Loop number of the loop we are annotating to.
        Loop invariants as strings.
        (optional) Loop assigns as strings.
        (optional) Loop decreases as strings.
        (optional) A symbol map that map symbol names in the loop contracts strings to their name in the symbol table of the goto model.

Loop contracts mangling consists of four steps.

    Construct a fake function with a fake loop that contains the loop contracts for parsing the contracts.
    Parse the fake function and extract the contracts as exprt.
    Substitute symbols in the extracted exprt with the symbols in the symbol table of the goto model according to the symbol map provided by the user.
    Typecheck all contract exprt.
    Annotate all contract exprt to the corresponding loop.
